### PR TITLE
Adding new "VIEW_MEMBERS" permissions restrict the scope of users visibility

### PR DESCRIPTION
**Pull request by jespino**
#### Summary
Adding new "VIEW_MEMBERS" permissions restrict the scope of users visibility

#### Ticket Link
[MM-14138](https://mattermost.atlassian.net/browse/MM-14138)

#### Checklist
- [x] Added or updated unit tests (required for all new features)
- [ ] Added API documentation (required for all new APIs)
- [ ] All new/modified APIs include changes to the drivers
- [x] [Has enterprise changes](https://github.com/mattermost/enterprise/pull/409)
- [ ] Has UI changes
- [ ] Includes text changes and localization file ([.../i18n/en.json](https://github.com/mattermost/mattermost-server/blob/master/i18n/en.json)) updates
- [ ] Touches critical sections of the codebase (auth, upgrade, etc.)

### CODE REVIEW

File: store/sqlstore/user_store.go [hunk 1]
```diff
@@ -1064,8 +1076,9 @@ func (us SqlUserStore) Count(options model.UserCountOptions) store.StoreChannel
 		}
 
 		if options.TeamId != "" {
-			query = query.LeftJoin("TeamMembers ON Users.Id = TeamMembers.UserId").Where("TeamMembers.TeamId = ? AND TeamMembers.DeleteAt = 0", options.TeamId)
+			query = query.LeftJoin("TeamMembers AS tm ON u.Id = tm.UserId").Where("tm.TeamId = ? AND tm.DeleteAt = 0", options.TeamId)
 		}
+		query = applyViewRestrictionsFilter(query, options.ViewRestrictions, false)
```
Comment (mgdelacroix) [inline store/sqlstore/user_store.go]: This is more a question than a change request. In this case, we can have two left joins for `TeamMembers` if `options.TeamId != ""`, is this ok for both PostgreSQL and MySQL?

File: api4/user.go [hunk 2]
```diff
@@ -111,14 +111,23 @@ func getUser(c *Context, w http.ResponseWriter, r *http.Request) {
...
-	// No permission check required
-
 	user, err := c.App.GetUser(c.Params.UserId)
 	if err != nil {
 		c.Err = err
 		return
 	}
 
+	canSee, err := c.App.UserCanSeeOtherUser(c.App.Session.UserId, c.Params.UserId)
+	if err != nil {
```
Comment (grundleborg) [inline api4/user.go]: This applies to many of these functions, but we need to check `UserCanSeeOtherUser` before calling `GetUser` as otherwise it's possible to determine whether a user exists even if you don't have permissions to view that user based on the error returned.

File: store/store.go [hunk 3]
```diff
@@ -256,12 +260,12 @@ type UserStore interface {
...
-	GetProfilesNotInChannel(teamId string, channelId string, offset int, limit int) StoreChannel
-	GetProfilesWithoutTeam(offset int, limit int) StoreChannel
-	GetProfilesByUsernames(usernames []string, teamId string) StoreChannel
+	GetProfilesNotInChannel(teamId string, channelId string, offset int, limit int, viewRestrictions *model.ViewUsersRestrictions) StoreChannel
+	GetProfilesWithoutTeam(offset int, limit int, viewRestrictions *model.ViewUsersRestrictions) StoreChannel
+	GetProfilesByUsernames(usernames []string, viewRestrictions *model.ViewUsersRestrictions) StoreChannel
 	GetAllProfiles(options *model.UserGetOptions) StoreChannel
 	GetProfiles(options *model.UserGetOptions) StoreChannel
-	GetProfileByIds(userId []string, allowFromCache bool) StoreChannel
+	GetProfileByIds(userId []string, allowFromCache bool, viewRestrictions *model.ViewUsersRestrictions) StoreChannel
```
Comment (grundleborg) [inline store/store.go]: Just want to check in general, but have you audited all the different stores and methods where this new permission might be applicable?

File: store/sqlstore/user_store.go (hunk 1 above)
Comment (jespino) [reply store/sqlstore/user_store.go]: Added a test and looks like it works well :)

File: store/store.go (hunk 3 above)
Comment (jespino) [reply store/store.go]: I tried, but there is a lot of places, I want to do a second pass before merging anyway, and add more tests to the API.

File: api4/user.go (hunk 2 above)
Comment (jespino) [reply api4/user.go]: Good point

File: app/user.go [hunk 4]
```diff
@@ -1845,3 +1869,142 @@ func (a *App) UpdateOAuthUserAttrs(userData io.Reader, user *model.User, provide
...
+	if userChannelMembers.Err != nil {
+		return nil, userChannelMembers.Err
+	}
+
+	channelIds := []string{}
+	for channelId := range userChannelMembers.Data.(map[string]string) {
+		channelIds = append(channelIds, channelId)
+	}
+
+	return &model.ViewUsersRestrictions{Teams: teamIdsWithPermission, Channels: channelIds}, nil
```
[the remark this answers is not in the document]
Comment (jespino) [reply app/user.go]: Yes, that is the assumption. Otherwise you wont be able almost to use the application (except for bots, for example)

File: app/user.go [hunk 5]
```diff
@@ -1719,12 +1723,24 @@ func (a *App) AutocompleteUsersInChannel(teamId string, channelId string, term s
...
 			return nil, err
 		}
-		uchan = a.Srv.Store.User().GetProfileByIds(uchanIds, false)
-		nuchan = a.Srv.Store.User().GetProfileByIds(nuchanIds, false)
+		if len(listOfAllowedChannels) == 0 {
+			return &model.UserAutocompleteInChannel{}, nil
+		}
+		searchChannelId := channelId
+		if !strings.Contains(strings.Join(listOfAllowedChannels, "."), channelId) {
+			searchChannelId = "INVALID-CHANNEL-ID"
```
[the remark this answers is not in the document]
Comment (jespino) [reply app/user.go]: This is a small trick to exclude the current channel if I haven't the permission to see the members of that channel. Was written with @mgdelacroix to avoid the limitations of the current elastic search implementation approach.

File: app/user.go [hunk 6]
```diff
@@ -1845,3 +1869,142 @@ func (a *App) UpdateOAuthUserAttrs(userData io.Reader, user *model.User, provide
...
+	for _, teamId := range teamIds {
+		if a.HasPermissionToTeam(userId, teamId, model.PERMISSION_VIEW_MEMBERS) {
+			teamIdsWithPermission = append(teamIdsWithPermission, teamId)
+		} else {
+			teamIdsWithoutPermission = append(teamIdsWithoutPermission, teamId)
+		}
+	}
+
+	if len(teamIdsWithoutPermission) == 0 {
+		return &model.ViewUsersRestrictions{Teams: teamIdsWithPermission}, nil
```
[the remark this answers is not in the document]
Comment (jespino) [reply app/user.go]: To build the restrictions you can have allowed Teams and allowed Channels, if you have only allowed teams, I don't build the allowedChannels at all (was an optimization, but maybe is not needed, and I think can have some corner cases). Actually because the cache, probably I don't need this optimization at all. I'll test it in the load tests.

File: store/sqlstore/team_store.go [hunk 7]
```diff
@@ -912,6 +926,22 @@ func (s SqlTeamStore) ResetAllTeamSchemes() store.StoreChannel {
 	})
 }
 
+var allTeamIdsForUserCache = utils.NewLru(ALL_TEAM_IDS_FOR_USER_CACHE_SIZE)
```
[the remark this answers is not in the document]
Comment (jespino) [reply store/sqlstore/team_store.go]: I have to check it, but I've been using the same approach that have the channel store caches.

File: app/user.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (mgdelacroix) [reply app/user.go]: I've been thinking about this and with the final interface, to avoid using this trick we can do:

```
uchanIds := []string{}
nuchanIds := []string{}
if !strings.Contains(strings.Join(listOfAllowedChannels, "."), channelId) {
	nuchanIds, err = a.Elasticsearch.SearchUsersInChannels(listOfAllowedChannels, term, options)
} else {
	uchanIds, nuchanIds, err = a.Elasticsearch.SearchUsersInChannelAndOutside(searchChannelId, listOfAllowedChannels, term, options)
}
if err != nil {
	return nil, err
}

...
```

It's a little longer but states more clearly the idea of "if your channel is not allowed, the autocompletion for channel users is an empty slice"

[the remark this answers is not in the document]
Comment (jespino) [reply app/user.go]: Good point, I'll do it :) Thanks

[the remark this answers is not in the document]
Comment (jespino) [reply app/user.go]: Done :)

File: api4/user.go (hunk 2 above)
Comment (jespino) [reply api4/user.go]: I think for now we can live it as is now, because anyway it is an improvement over the previous situation (no restriction at all). But yes, probably we should have some kind of policy here like return always the same error for both cases.

File: store/sqlstore/team_store.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (jespino) [reply store/sqlstore/team_store.go]: Ok, is not syncrhonised across the cluster, so, you can have the posibility of have it inconsistent for 30 minutes (the TTL of the cache). I going to down it into 15 min, but it was replicated from channel store, so the channel store have the same problem (I guess).

[the remark this answers is not in the document]
Comment (jespino) [reply store/sqlstore/team_store.go]: Ok, I found finally the code that I wasn't seing.

[the remark this answers is not in the document]
Comment (jespino) [reply store/sqlstore/team_store.go]: Done! :)

File: app/user.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (grundleborg) [reply app/user.go]: OK makes sense.

File: api4/user.go (hunk 2 above)
Comment (grundleborg) [reply api4/user.go]: Please can you move the `UserCanSeeOtherUser` call to *before* the `GetUser` call?

Comment (jespino) [reply api4/user.go]: Sure

Comment (jespino) [reply api4/user.go]: But there are other cases that Is more complex, when you try to get the user by username or by email, where I have to try to get the user first (not return the error anyway, and check if I can see the user), and then return the error depending on it.

**Comment by amyblais** [comment]
A quick reminder that tomorrow (v5.10 Feature Complete) is the last day to merge feature PRs if we want this for v5.10.

File: api4/user.go (hunk 2 above)
Comment (grundleborg) [reply api4/user.go]: The issue is that if we ignore this, we'll get a bug bounty security ticket against it, so I think we have to get it right now, even if it's a bit complicated.

Comment (jespino) [reply api4/user.go]: It is already done

File: store/store.go (hunk 3 above)
Comment (jespino) [reply store/store.go]: I'm working on this and I found 2 more cases that I need to control (team members), almost done, and user statuses quite easy. I'll submit a commit for that soon.

File: api4/status.go [hunk 8]
```diff
@@ -45,7 +54,18 @@ func getUserStatusesByIds(c *Context, w http.ResponseWriter, r *http.Request) {
...
-	// No permission check required
+	visibleUserIds := []string{}
+	for _, userId := range userIds {
+		canSee, err := c.App.UserCanSeeOtherUser(c.App.Session.UserId, c.Params.UserId)
+		if err == nil && canSee {
+			visibleUserIds = append(visibleUserIds, userId)
+		}
+	}
+
+	if len(userIds) == 0 {
```
Comment (grundleborg) [inline api4/status.go]: Shouldn't you be using `visibleUserIds` here and below?

File: store/sqlstore/user_store.go [hunk 9]
```diff
@@ -1039,18 +1051,18 @@ func (us SqlUserStore) PermanentDelete(userId string) store.StoreChannel {
 
 func (us SqlUserStore) Count(options model.UserCountOptions) store.StoreChannel {
 	return store.Do(func(result *store.StoreResult) {
-		query := sq.Select("COUNT(Users.Id)").From("Users")
+		query := sq.Select("COUNT(DISTINCT u.Id)").From("Users AS u")
```
Comment (agnivade) [inline store/sqlstore/user_store.go:1054]: Hey @jespino - I ran into this code while doing something else. Curious to know if we need the "DISTINCT" function here? Id is already a primary key of the users table. And I don't see joins with the same table, or any unions of sorts.

Comment (jespino) [reply store/sqlstore/user_store.go:1054]: It is hard to try, but as part of the "ApplyViewRestrictions" function to the query, it joins with the team members and the channel members, so I guess the `DISTINCT` was added there to avoid duplicated rows based on that.

Comment (agnivade) [reply store/sqlstore/user_store.go:1054]: Hmm .. okay but you are joining with the user id. So there shouldn't be any duplicated rows right?
